multipass: init at 1.11.0

jnsgruk opened the pull request:
###### Description of changes

Adds a [multipass](https://multipass.run) package and NixOS module.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] ~~aarch64-linux~~
  - [ ] ~~x86_64-darwin~~
  - [ ] ~~aarch64-darwin~~
- [ ] ~~For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))~~
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jnsgruk committed:
maintainers: add jnsgruk

AndersonTorres commented on nixos/modules/virtualisation/multipass.nix:
```diff
@@ -0,0 +1,74 @@
+    environment.systemPackages = [cfg.package];
+
+    systemd.services.multipass = {
+      description = "Multipass orchestrates virtual Ubuntu instances";
```
```suggestion
      description = "Multipass orchestrates virtual Ubuntu instances.";
```

AndersonTorres commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,74 @@
+  version = "1.11.0";
+in
+  stdenv.mkDerivation {
+    name = "${pname}-${version}";
```
`name` was deprecated years ago. Just inherit them:


```suggestion
    inherit pname version;
```

AndersonTorres commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,74 @@
+
+    meta = with lib; {
+      description = "A tool that orchestrates virtual Ubuntu instances";
+      homepage = "https://multipass.run";
```
Deprecated years ago.

```suggestion
      homepage = "https://multipass.run";
```

AndersonTorres commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,74 @@
+    meta = with lib; {
+      description = "A tool that orchestrates virtual Ubuntu instances";
+      homepage = "https://multipass.run";
+      license = licenses.gpl3;
```
```suggestion
      license = licenses.gpl3Plus;
```

AndersonTorres reviewed (changes requested):
General hammering.

jnsgruk commented on nixos/modules/virtualisation/multipass.nix:
```diff
@@ -0,0 +1,74 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
```
Got it - will address in a follow up commit. Thanks!

jnsgruk commented on nixos/modules/virtualisation/multipass.nix:
```diff
@@ -0,0 +1,74 @@
+  pkgs,
+  ...
+}:
+with lib; let
```
Got it, thanks

jnsgruk commented:
Thanks for the rapid review @AndersonTorres! I think I've addressed all your feedback.

jnsgruk commented:
I've added some simple tests here; I initially added a test that actually started a VM and tested it, but since Multipass fetches the VM images from the internet, the tests don't pass, so I've removed them for now.

RaitoBezarius commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,93 @@
+      --replace 'set(MULTIPASS_VERSION ''${MULTIPASS_VERSION})' 'set(MULTIPASS_VERSION "v${version}")'
+
+    substituteInPlace ./src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp \
+      --replace "OVMF.fd" "${OVMF.fd}/FV/OVMF.fd"
```
Don't you need to depend on the actual targetPlatform here?
I introduced something like: `OVMF.firmware` so you don't have to think about it, it is only the code though.

jnsgruk replied:
Nice, I'll do that!

RaitoBezarius commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,93 @@
+
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=Release"
+    "-DMULTIPASS_ENABLE_TESTS=off"
```
Can you add a comment on why the tests have to be disabled?

jnsgruk replied:
I was struggling before with Cmake going to fetch googletest from the internet; working on a patch for this now!

RaitoBezarius replied:
Awesome! <3

RaitoBezarius commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,93 @@
+  ];
+
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=Release"
```
I'm surprised this needs to be passed at all, do you think it didn't work on the first try without it?

jnsgruk replied:
This will be removed in the next commit!

RaitoBezarius commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,93 @@
+      dnsmasq
+      iproute2
+      iptables
+      OVMF.fd
```
`OVMF.firmware`

jnsgruk replied:
Nice, thanks :)

RaitoBezarius commented:
> I've added some simple tests here; I initially added a test that actually started a VM and tested it, but since Multipass fetches the VM images from the internet, the tests don't pass, so I've removed them for now.

Thank you a lot for this! Can you rebase your branch into three commits:
- maintainer introduction commit
- multipass: init at 1.11.0
- nixos/multipass: init

separating package and module if possible :). (you can keep tests in the module commit, if it's too annoying, make it two commits with `multipass: init at 1.11.0`

jnsgruk commented:
> > I've added some simple tests here; I initially added a test that actually started a VM and tested it, but since Multipass fetches the VM images from the internet, the tests don't pass, so I've removed them for now.
> 
> Thank you a lot for this! Can you rebase your branch into three commits:
> 
> * maintainer introduction commit
> * multipass: init at 1.11.0
> * nixos/multipass: init
> 
> separating package and module if possible :). (you can keep tests in the module commit, if it's too annoying, make it two commits with `multipass: init at 1.11.0`

Will address your feedback, then rebase. Thanks!

jnsgruk commented:
> > I've added some simple tests here; I initially added a test that actually started a VM and tested it, but since Multipass fetches the VM images from the internet, the tests don't pass, so I've removed them for now.
> 
> Thank you a lot for this! Can you rebase your branch into three commits:
> 
> * maintainer introduction commit
> * multipass: init at 1.11.0
> * nixos/multipass: init
> 
> separating package and module if possible :). (you can keep tests in the module commit, if it's too annoying, make it two commits with `multipass: init at 1.11.0`

Will address your feedback, then rebase. Thanks!

RaitoBezarius commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,131 @@
+  };
+
+  meta = with lib; {
+    description = "Ubuntu VMs on demand for any worktation.";
```
```suggestion
    description = "Ubuntu VMs on demand for any workstation.";
```

AndersonTorres commented on nixos/modules/virtualisation/multipass.nix:
```diff
@@ -0,0 +1,76 @@
+  cfg = config.virtualisation.multipass;
+in
+{
+  options = with lib; {
```
```suggestion
  options = {
```

AndersonTorres replied:
As said before, don't use `with lib;`.

AndersonTorres commented:
> you can keep tests in the module commit

It is recommended to put the tests in its own separated commit.

jnsgruk commented on nixos/modules/virtualisation/multipass.nix:
```diff
@@ -0,0 +1,76 @@
+{
+  options = with lib; {
+    virtualisation.multipass = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = mdDoc ''
+          This option enables multipass, a daemon that manages
+          virtualised Ubuntu instances with a simple command
+          line interface.
+        '';
+      };
```
Thanks! Will change this.

jnsgruk commented on nixos/modules/virtualisation/multipass.nix:
```diff
@@ -0,0 +1,76 @@
+      };
+
+      logLevel = mkOption {
+        type = types.str;
```
Nice, thanks!

jnsgruk commented on nixos/modules/virtualisation/multipass.nix:
```diff
@@ -0,0 +1,76 @@
+        '';
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.multipass;
+        defaultText = literalExpression "pkgs.multipass";
+        description = mdDoc ''
+          The multipass package to use.
+        '';
+      };
```
TIL!

jnsgruk committed:
multipass: init at 1.11.0

jnsgruk committed:
nixos/multipass: init

jnsgruk committed:
nixos/tests/multipass: init

jnsgruk commented on nixos/tests/multipass.nix:
```diff
@@ -0,0 +1,42 @@
+      documentation.enable = lib.mkForce false;
+
+      # Our tests require `grep` & friends:
+      environment.systemPackages = with pkgs; [ busybox ];
```
Fixed, this was actually not required at all any more.

jnsgruk commented:
I've restricted this to `x86_64-linux` for now, as I don't have a way of testing others. I know for sure there will be tweaks needed in the build for `aarch64`.

AndersonTorres commented on nixos/tests/multipass.nix:
```diff
@@ -0,0 +1,39 @@
+{
+  name = "multipass";
+
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ jnsgruk ];
+  };
```
```suggestion
  meta.maintainers = [ lib.maintainers.jnsgruk ];
```

jnsgruk replied:
Fixed! Thanks again for all the feedback!

jnsgruk replied:
Fixed! Thanks again for all the feedback!

AndersonTorres reviewed (changes requested):
Besides LGTM

jnsgruk commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,128 @@
+    substituteInPlace ./tests/CMakeLists.txt \
+      --replace "FetchContent_MakeAvailable(googletest)" ""
+
+    cat >> tests/CMakeLists.txt <<'EOF'
```
It could be, but this is taken care of automatically by `FetchContent_MakeAvailable` in the upstream I believe. That doesn't work here as it relies on fetching `googletest` from the internet - and thus fails in the nix build. This is why originally I was building without the tests - the upstream tests work fine in build environments that can reach the internet.

townsend2010 replied:
As an upstream maintainer of Multipass, we are indeed "living at the head" per [Google's recommendation](https://github.com/google/googletest#live-at-head) and thus need to keep it the way it currently is.

Thanks for this!

AndersonTorres commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,128 @@
+      --replace "QEMU_EFI.fd" "${OVMF.fd}/FV/QEMU_EFI.fd"
+  '';
+
+  patchPhase = ''
```
I almost forgot this!

```suggestion
  postPatch = ''
```

jnsgruk replied:
Fixed!

AndersonTorres reviewed (changes requested):
Besides LGTM.

AndersonTorres commented:
And, before I forget: welcome to Nixpkgs!

jnsgruk commented:
Thanks for all the help! 😄

SuperSandro2000 commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,128 @@
+    repo = "multipass";
+    rev = "refs/tags/v${version}";
+    sha256 = "sha256-2d8piIIecoSI3BfOgAVlXl5P2UYDaNlxUgHXWbnSdkg=";
+    fetchSubmodules = true;
```
Using submodules is problematic because eg we are using a two year old fork of libssh https://github.com/canonical/multipass/blob/main/.gitmodules#L9-L11

RaitoBezarius replied:
Please, can you open an issue documenting your remarks post-merge?

SuperSandro2000 replied:
As this is directly related to this PR, I think that is unnecessary.

SuperSandro2000 commented on pkgs/tools/virtualization/multipass/default.nix:
```diff
@@ -0,0 +1,128 @@
+
+  nativeBuildInputs = [
+    cmake
+    git
```
Why do we need git?
